ace/snippets/python.js

```diff
@@ -1,50 +1,8 @@
):\n		"""${3:docstring for $1}"""\n		def __init__(self, ${4:arg}):\n			${5:super($1, self).__init__()}\n			self.$4 = $4\n			${6}\n# New Function\nsnippet def\n	def ${1:fname}(${2:`indent(\'.\') ? \'self\' : \'\'`}):\n		"""${3:docstring for $1}"""\n		${4:# TODO: write code...}\nsnippet deff\n	def ${1:fname}(${2:`indent(\'.\') ? \'self\' : \'\'`}):\n		${3:# TODO: write code...}\n# New Method\nsnippet defs\n	def ${1:mname}(self, ${2:arg}):\n		${3:# TODO: write code...}\n# New Property\nsnippet property\n	def ${1:foo}():\n		doc = "${2:The $1 property.}"\n		def fget(self):\n			${3:return self._$1}\n		def fset(self, value):\n			${4:self._$1 = value}\n# Ifs\nsnippet if\n	if ${1:condition}:\n		${2:# TODO: write code...}\nsnippet el\n	else:\n		${1:# TODO: write code...}\nsnippet ei\n	elif ${1:condition}:\n		${2:# TODO: write code...}\n# For\nsnippet for\n	for ${1:item} in ${2:items}:\n		${3:# TODO: write code...}\n# Encodes\nsnippet cutf8\n	# -*- coding: utf-8 -*-\nsnippet clatin1\n	# -*- coding: latin-1 -*-\nsnippet cascii\n	# -*- coding: ascii -*-\n# Lambda\nsnippet ld\n	${1:var} = lambda ${2:vars} : ${3:action}\nsnippet .\n	self.\nsnippet try Try/Except\n	try:\n		${1:# TODO: write code...}\n	except ${2:Exception}, ${3:e}:\n		${4:raise $3}\nsnippet try Try/Except/Else\n	try:\n		${1:# TODO: write code...}\n	except ${2:Exception}, ${3:e}:\n		${4:raise $3}\n	else:\n		${5:# TODO: write code...}\nsnippet try Try/Except/Finally\n	try:\n		${1:# TODO: write code...}\n	except ${2:Exception}, ${3:e}:\n		${4:raise $3}\n	finally:\n		${5:# TODO: write code...}\nsnippet try Try/Except/Else/Finally\n	try:\n		${1:# TODO: write code...}\n	except ${2:Exception}, ${3:e}:\n		${4:raise $3}\n	else:\n		${5:# TODO: write code...}\n	finally:\n		${6:# TODO: write code...}\n# if __name__ == \'__main__\':\nsnippet ifmain\n	if __name__ == \'__main__\':\n		${1:main()}\n# __magic__\nsnippet _\n	__${1:init}__${2}\n# python debugger (pdb)\nsnippet pdb\n	import pdb; pdb.set_trace()\n# ipython debugger (ipdb)\nsnippet ipdb\n	import ipdb; ipdb.set_trace()\n# ipython debugger (pdbbb)\nsnippet pdbbb\n	import pdbpp; pdbpp.set_trace()\nsnippet pprint\n	import pprint; pprint.pprint(${1})${2}\nsnippet "\n	"""\n	${1:doc}\n	"""\n# test function/method\nsnippet test\n	def test_${1:description}(${2:self}):\n		${3:# TODO: write code...}\n# test case\nsnippet testcase\n	class ${1:ExampleCase}(unittest.TestCase):\n		\n		def test_${2:description}(self):\n			${3:# TODO: write code...}\nsnippet fut\n	from __future__ import ${1}\n#getopt\nsnippet getopt\n	try:\n		# Short option syntax: "hv:"\n		# Long option syntax: "help" or "verbose="\n		opts, args = getopt.getopt(sys.argv[1:], "${1:short_options}", [${2:long_options}])\n	\n	except getopt.GetoptError, err:\n		# Print debug info\n		print str(err)\n		${3:error_action}\n\n	for option, argument in opts:\n		if option in ("-h", "--help"):\n			${4}\n		elif option in ("-v", "--verbose"):\n			verbose = argument\n',t.scope="python"});                (function() {
```

Comment:
I think this file was edited to have custom snippets in https://github.com/bbcmicrobit/PythonEditor/commit/4ee0e78d20c39cd58a578c823057f5fd5e177768, as this is what it look like now:

![image](https://user-images.githubusercontent.com/29712657/54597866-3f741000-4a2f-11e9-8dca-e7cb8ab25994.png)

To avoid this issue in the future if ACE is updated again, can you check if you can move the old snippets to a file like 'python_microbit.js' and then change how we load them. Hopefully it's just these lines:


https://github.com/bbcmicrobit/PythonEditor/blob/07a8403d21d54e00ad6fb11534e8b65e60153e0f/python-main.js#L53-L56

Reply:
There are a lot of unused files (snippets, themes, mode files for other languages). Should we remove these in a separate issue?

Reply:
Do they all come together as part of a bundle? If so we might as well leave them, as only the used files are loaded into the editor.

Reply:
https://github.com/ajaxorg/ace-builds/tree/master/src-min

Reply:
Okay, so let's keep them in this case

Reply:
It appears to be a bit tricky to split out this change as the snippets appear to be tied to the mode. What's the best approach here?

Reply:
I'm having a look now to see if there is anything we can monkeypatch or inject, otherwise we'll replace the original as done previously.

Reply:
Just to give you an update, we could bypass the python snippets by generating our own and then using something like `snippetManager.parseSnippetFile()` to create the object that Ace needs.
Unfortunately I'm having trouble getting the right output from `snippetManager.parseSnippetFile()` when I run the code in the debugger _sometimes_ I get the correct object generated, and most of the time I get an empty array, not sure what's going on.

Reply:
After hitting a couple of walls I was able to add custom snippets in commit feb671db42c009a00b3868af82818e568d41f2c1 in this branch.

Reply:
Have a play to make sure all works as expected, if it's all good then we should keep the Ace upstream python.js file and rely on our own snippets for the editor.

Reply:
LGTM, nice one.